src/server.py

```diff
@@ -27,38 +35,130 @@ def server():
             if len(part) < buffer_length:
                 message_complete = True
         print(message)
-        conn.send(response_ok())
+
+        try:
+            uri = parse_request(message)
+        except HTTPException as exception:
+            msg = response_error(exception)
+        try:
+            file_data_tuple = resolve_uri(uri)
+        except HTTPException as exception:
+            msg = response_error(exception)
+        try:
+            msg = response_ok(file_data_tuple)
+        except HTTPException:
+            msg = (response_error("500 Internal Server Error"))
+
+        conn.sendall(msg.encode("utf-8"))
         conn.close()
 
 
-def response_ok():
-    """Function returns an HTTP '200 OK' response."""
-    response = (b"HTTP/1.1 200 OK\r\n"
-                b"Host: 127.0.0.1:5000\r\n\r\n"
-                b"Connection Successful")
-    return response
+def resolve_uri(uri):
+    if uri.find('../') is not -1:
+        raise HTTPException('403 Forbidden')
+    elif uri[-1] == '/':
+        return response_ok((generate_directory_html(uri), 'text/html'))
+    else:
+        try:
+            file_data_tuple = get_file_data(uri)
+        except HTTPException:
+            raise HTTPException('404 File Not Found')
+        return file_data_tuple
 
 
-def response_error():
-    """Function returns an HTTP '500 Internal Server Error' response."""
-    response = (b"HTTP/1.1 500 Internal Server Error\r\n"
-                b"Host: 127.0.0.1:5000\r\n\r\n"
-                b"Server Error")
-    return response
+def get_file_data(uri):
+    '''Reads file and returns a tuple: (file_dat, content-type).'''
+    try:
+        file_b = io.open(ROOT + '/' + uri, 'rb')
+    except IOError:
+        raise HTTPException('404 File Not Found')
```

Comment:
whoa, wait a moment. You *do* create the right type of HTTPException here.  Why are you catching and raising a new one in the `resolve_uri` function where this function is called.  

Remember that if your function does not catch an exception, it will continue to bubble up until something does catch it, or until it reaches the end of the call stack and kills the running process as an unhandled exception.